hack/generator/pkg/codegen/pipeline_apply_kubernetes_resource_interface.go

```diff
@@ -19,32 +20,31 @@ func applyKubernetesResourceInterface(idFactory astmodel.IdentifierFactory) Pipe
 		"Ensures that every resource implements the KubernetesResource interface",
 		func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
 
+			skip := make(map[astmodel.TypeName]struct{})
+
 			result := make(astmodel.Types)
-			for _, typeDef := range types {
+			for typeName, typeDef := range types {
+				if _, ok := skip[typeName]; ok {
+					continue
+				}
 
 				if resource, ok := typeDef.Type().(*astmodel.ResourceType); ok {
-
-					specName, ok := resource.SpecType().(astmodel.TypeName)
-					if !ok {
-						return nil, errors.Errorf("resource %q spec was not of type TypeName, instead: %T", typeDef.Name(), typeDef.Type())
-					}
-
-					spec, ok := types[specName]
-					if !ok {
-						return nil, errors.Errorf("couldn't find resource spec %q", specName)
-					}
-
-					specObj, err := astmodel.TypeAsObjectType(spec.Type())
+					newResource, err := resource.WithKubernetesResourceInterfaceImpl(idFactory, types)
 					if err != nil {
-						return nil, errors.Wrapf(err, "resource spec %q did not contain an object", specName)
+						return nil, errors.Wrapf(err, "Couldn't implement Kubernetes resource interface for %q", typeName)
 					}
 
-					iface, err := astmodel.NewKubernetesResourceInterfaceImpl(idFactory, specObj)
-					if err != nil {
-						return nil, errors.Wrapf(err, "Couldn't implement Kubernetes resource interface for %q", spec.Name())
+					// this is really very ugly; a better way?
```

Comment:
You could make `WithKubernetesResourceInterfaceImpl` return a triplet with one return being an optional updated `SpecType`. It makes it a bit more obvious from the function signature that there's a possible modification? Or if the triplet result is too ugly you could make a little struct for it.

It's honestly not _lots_ better but might be a bit better.